Fix several overflows in box and track processing

oftheforest opened the pull request.

oftheforest committed:
Fix several overflows in box and track processing

jessa0 commented on src/mp4box/co64.rs:
```diff
@@ -49,6 +49,11 @@ impl<R: Read + Seek> ReadBox<&mut R> for Co64Box {
         let (version, flags) = read_box_header_ext(reader)?;
 
         let entry_count = reader.read_u32::<BigEndian>()?;
+        if u64::from(entry_count) > size.saturating_sub(4) / 8 {
```
One initial thought which applies to most of this PR: typically, I try to either write code where magic numbers are documented implicitly using a constant or variable name or explicitly in a comment. An example for this line:

```Rust
let box_header_size = mem::size_of::<u32>();
let entry_size = mem::size_of::<u64>();
if u64::from(entry_count) > size.saturating_sub(box_header_size) / entry_size {
```

Now, I realize that the author of this library hasn't exactly taken this approach elsewhere.. so I'm not sure what they prefer. OTOH, the PR is a bit difficult to review as-is, since I have to make these calculations myself in order to verify them. I'll leave it up to you, considering the above, but what do you think?

jessa0 replied:
Another idea to make these types of comparisons more readable:

```Rust
let box_header_size = mem::size_of::<u32>();
let entry_size = mem::size_of::<u64>();
let possible_entries_size = size.saturating_sub(box_header_size);
if u64::from(entry_count) > possible_entries_size / entry_size {
```

jessa0 replied:
fixed in [df4e2a7](https://github.com/privacyresearchgroup/mp4-rust/pull/2/commits/df4e2a72af19b7388691830f68834fa63be848b5)

oftheforest committed:
Use size_of::<Type>() instead of magic numbers

oftheforest committed:
Fix a panic in Mp4Track::read_sample() for one-past-the-end

This appears to be a bug unmasked by other changes. read_sample() calls
sample_offset() then sample_size(), and assumes that if the former returns Ok
then the latter does as well. However, if the sample_id is one past the end,
sample_offset() might succeed (it only checks samples _up to_ the given
sample_id but not _including_ it) while sample_size() fails (because the sample
doesn't exist). read_sample() will then panic.

Fix this by duplicating the error propagation (that is currently done for
sample_offset) for sample_size, instead of unwrapping. This is a cautious
change that fixes the bug; alternatively, having sample_offset() call
sample_size() on the given sample_id and propagate any error might also work.

jessa0 commented on src/mp4box/elst.rs:
```diff
@@ -63,6 +64,18 @@ impl<R: Read + Seek> ReadBox<&mut R> for ElstBox {
+            size_of::<u64>() + size_of::<i64>() // segment_duration + media_time
+        } else {
+            size_of::<u32>() + size_of::<i32>() // segment_duration + media_time
+        } + size_of::<i16>()
```
This might be a nit, but I think the "if statement as an expression" makes rustfmt go a bit nuts and is a little confusing to read. Here's a suggestion:

```Rust
let entry_size = {
    let mut entry_size = if version == 1 {
        size_of::<u64>() + size_of::<i64>() // segment_duration + media_time
    } else {
        size_of::<u32>() + size_of::<i32>() // segment_duration + media_time
    };
    entry_size += size_of::<i16>() + size_of::<i16>(); // media_rate_integer + media_rate_fraction
    entry_size
};
```

Though, it might be better at that point to factor it out into a separate method..

jessa0 replied:
fixed in [754ea1c](https://github.com/privacyresearchgroup/mp4-rust/pull/2/commits/754ea1ce9d967aa241f6d2b1e9175db23a500eca)

jessa0 commented on src/mp4box/co64.rs:
```diff
@@ -48,7 +49,14 @@ impl<R: Read + Seek> ReadBox<&mut R> for Co64Box {
 
         let (version, flags) = read_box_header_ext(reader)?;
 
+        let other_size = size_of::<u32>(); // entry_count
```
Hmm, should `other_size` also include the amount read by `read_box_header_ext` (which I believe is 4 bytes?). There are a bunch of instances of this below as well.

oftheforest replied:
Good catch, thanks! I missed that the `size` argument to `read_box()` is actually the total box size (_not_ subtracting size/type/version/flags).

jessa0 replied:
It's not exactly, but you can kinda treat it like that. In the case the optional "64-bit size" box header field is present, the `size` parameter will actually be the total box size (including box header) _minus 8_. I believe this is a hack, and it's hard to grok how it all still works properly together.

The long story short is that you can basically count on the `size` parameter being the length of the box data (i.e. everything but the box header) plus 8.

jessa0 replied:
looks addressed by [1c8ca74](https://github.com/privacyresearchgroup/mp4-rust/pull/2/commits/1c8ca74b93664391d1c1441ec45bf36b6b74e690)

jessa0 replied:
fixed in [1c8ca74](https://github.com/privacyresearchgroup/mp4-rust/pull/2/commits/1c8ca74b93664391d1c1441ec45bf36b6b74e690)

jessa0 commented on src/mp4box/stts.rs:
```diff
@@ -55,6 +55,11 @@ impl<R: Read + Seek> ReadBox<&mut R> for SttsBox {
         let (version, flags) = read_box_header_ext(reader)?;
 
         let entry_count = reader.read_u32::<BigEndian>()?;
+        if u64::from(entry_count) > size.saturating_sub(4) / 8 {
```
Still need to document these.

jessa0 replied:
looks addressed by [1c8ca74](https://github.com/privacyresearchgroup/mp4-rust/pull/2/commits/1c8ca74b93664391d1c1441ec45bf36b6b74e690)

jessa0 commented on src/mp4box/trun.rs:
```diff
@@ -83,6 +84,16 @@ impl<R: Read + Seek> ReadBox<&mut R> for TrunBox {
+            + if TrunBox::FLAG_SAMPLE_SIZE & flags > 0 { size_of::<u32>() } else { 0 } // sample_size
+            + if TrunBox::FLAG_SAMPLE_FLAGS & flags > 0 { size_of::<u32>() } else { 0 } // sample_flags
+            + if TrunBox::FLAG_SAMPLE_CTS & flags > 0 {
+                if version == 0 { size_of::<u32>() } else { size_of::<i32>() } // sample_composition_time_offset
```
Can probably just collapse this into `size_of::<u32>()`.

jessa0 replied:
fixed in [619a36b](https://github.com/privacyresearchgroup/mp4-rust/pull/2/commits/619a36b38b435b09ba997434d810f0eab2ad3b35)

jessa0 commented on src/mp4box/trun.rs:
```diff
@@ -97,10 +108,30 @@ impl<R: Read + Seek> ReadBox<&mut R> for TrunBox {
+        let mut sample_sizes = Vec::new();
+        let mut sample_flags = Vec::new();
+        let mut sample_cts = Vec::new();
+        if u64::from(sample_count) * u64::from(sample_size as u64)
```
There's a no-op conversion here: `u64::from(sample_size as u64)`.

jessa0 replied:
fixed in [619a36b](https://github.com/privacyresearchgroup/mp4-rust/pull/2/commits/619a36b38b435b09ba997434d810f0eab2ad3b35)

oftheforest committed:
Account for header size in box processing overflow fixes

jessa0 reviewed (approved):
LGTM now!

oftheforest committed:
Ensure that boxes aren't bigger than their containers

Together with the entry_count checks, this eliminates several OOMs when reading
incorrect mp4 files.

jessa0 commented on src/mp4box/avc1.rs:
```diff
@@ -103,6 +103,11 @@ impl<R: Read + Seek> ReadBox<&mut R> for Avc1Box {
 
         let header = BoxHeader::read(reader)?;
         let BoxHeader { name, size: s } = header;
+        if s > size {
```
I suppose we could be more precise here (and below), with:

```Rust
if s > size.saturating_sub(reader.stream_position() - start) {
```

Do you think it's worth it to do so? Maybe not, since I think these checks are mostly to prevent crashes due to large allocations, right? In that case, maybe add comments explaining why the checks are imprecise.

This also makes me realize that pretty much every `read_box` should be doing:

```Rust
let mut reader = reader.take(size - HEADER_SIZE);
```

so that they don't read past the end of the box. But I can file a bug (#4) for that to do in a separate PR (since this one is getting large).

oftheforest replied:
I seem to remember that this/something else already effectively prevented that sort of box overrun? I'll take another look, though; if that's not the case, I'll open another PR for #4.

oftheforest committed:
Fix order of arithmetic operations

This was due to an incorrect transcription when switching to checked
arithmetic, and fixes a bug that could cause attempted lookups of the wrong
chunk_id.

jessa0 reviewed (approved):
LGTM!
